File: tests/matrix/targets/macho/fix.go
// Copyright 2026 The Sismo Authors. All rights reserved.
// Licensed under the MIT License.
//
// Go port of fix.c: markers land in the symtab as main.sismoFixLeaf etc.,
// with line info carried by .gopclntab rather than a debug map.

package main

import "fmt"

//go:noinline
func sismoFixLeaf(x int) int {
	acc := 0
	for i := 0; i < x; i++ {
		acc += i * i
	}
	return acc
}

//go:noinline
func sismoFixMid(x int) int {
	return sismoFixLeaf(x) + 1
}

func main() {
	fmt.Println(sismoFixMid(100))
}

File: tests/matrix/targets/macho/fix.zig
// Copyright 2026 The Sismo Authors. All rights reserved.
// Licensed under the MIT License.
//
// Zig port of fix.c: zig's linker embeds DWARF straight into the binary
// (no debug map), and markers land in the symtab as fix.sismoFixLeaf etc.

fn sismoFixLeaf(x: u64) u64 {
    var acc: u64 = 0;
    var i: u64 = 0;
    while (i < x) : (i += 1) acc +%= i *% i;
    return acc;
}

fn sismoFixMid(x: u64) u64 {
    return @call(.never_inline, sismoFixLeaf, .{x}) + 1;
}

var seed: u64 = 100;

pub fn main() void {
    // The volatile load keeps the input runtime-only: ReleaseFast otherwise
    // const-folds the whole marker chain out of the binary.
    const p: *volatile u64 = &seed;
    const v = @call(.never_inline, sismoFixMid, .{p.*});
    std.mem.doNotOptimizeAway(v);
}

const std = @import("std");
